[WIP][CI] Enable clang v3.8 on Travis CI

PeterDaveHello opened the pull request:
- Clean up clang related codes and config
- Remove lldb-3.8 package since we don't need, though it's on llvm install instructions.
- The origin codes have a bug about the PATH which makes clang would not be detected/used correctly.
- Remove clang version print.
- Remove useless gcc/g++ related packages and variables.

ljharb commented on .travis.yml:
```diff
@@ -30,45 +30,12 @@ script:
     - SHELLCHECK=true
     - SHELL=bash TEST_SUITE=install_script
     - SHELL=sh TEST_SUITE=fast
-    - SHELL=dash TEST_SUITE=fast
```
Um, what?

ljharb replied:
You can enable your own repo on Travis and push branches to it without creating PRs, btw

PeterDaveHello replied:
@ljharb yes I just did, that's what I did before that, and just found the conflicts with #1311, so just wait for that, sorry about this.

PeterDaveHello committed:
[CI] Clean up clang related codes and config

- Remove lldb-3.8 package since we don't need, though it's on llvm
instructions.
- Fix clang PATH bug which makes clang would not be detected/used correctly.
- Remove clang version print.

PeterDaveHello committed:
[CI] Remove gcc/g++ related pkgs & vars which are useless now.

PeterDaveHello commented:
@ljharb I fixed the conflicts 😄

ljharb commented on .travis.yml:
```diff
@@ -24,10 +21,8 @@ before_install:
-  - clang++ --version
   - if [ -n "${SHELLCHECK-}" ]; then stack setup && stack install ShellCheck && shellcheck --version ; fi
-  - if [ -z "${SHELLCHECK-}" ]; then sudo ln -sf /usr/bin/clang-3.8 /usr/bin/clang && sudo ln -sf /usr/bin/clang++-3.8 /usr/bin/clang++ && clang --version ; fi
+  - mkdir -p $HOME/bin && ln -s /usr/bin/clang-3.8 $HOME/bin/clang && ln -s /usr/bin/clang++-3.8 $HOME/bin/clang++
```
please surround all the paths in quotes, in case `$HOME` has a space in it.

ljharb commented:
Hmm - my main concern here is that now we won't be testing gcc-based installation.

Test run speed is less important than testing the maximum amount of use cases.

PeterDaveHello commented:
@ljharb to be honest, I don't think we need to test nodejs as its author will test it, on both gcc and clang, as they write the doc about compilation using gcc/clang, that's their job and we can do nothing, what we should do here is to make sure we pass the correct parameters to "make", am I right?

ljharb commented:
The thing that needs testing is how *nvm* compiles it - not that node itself compiles, as it surely will.

While it's a very fair point that simply asserting that we pass the correct parameters to `make` is a *great* unit test - and we *definitely should* add "fast" unit tests for `nvm install -s` that stubs out `nvm_download_artifact` and `make`; that's a wonderful idea - but what that fails to do is *integration test* `nvm install -s` on `gcc`, which is the purpose of the test.

PeterDaveHello commented:
So maybe just separate the "make" part to an individual test for both gcc and clang? Then We don't need to "make" nodejs so many times per CI round.

ljharb commented:
Yes, I think that would be best - in other words:
 1. run at least one `nvm install -s` test with gcc
 1. run at least one `nvm install -s` test with clang
 1. make unit tests that stub things out and test calls to `make` for gcc
 1. make unit tests that stub things out and test calls to `make` for clang

That should cover all the testing, without adding any slow tests to the build.

PeterDaveHello commented:
@ljharb what's the parameters should be tested in that? If we are not passing some special cases with potential problems, I really doubt if we need to build nodejs in the test.

ljharb commented:
In numbers 3 and 4? We should test all the permutations we're likely to produce, but because `make` will be stubbed out, we won't need to build node at all for those (just mock out some of the filesystem stuff that the code expects)

PeterDaveHello commented:
I mean in number 1 and 2, as we won't bump the nodejs version to be tested automatically, a certain build with the *same parameters* has built successfully should always build successfully with the *same parameters*, can't we just mock this part by hard code?

ljharb commented:
Sure, unless the tarballs change retroactively (as they have before).

vincent89291 reviewed (commented):
Explicit
